src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -0,0 +1,397 @@
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+import base64
+import copy
+import json
+import logging
+import pathlib
+import sys
+from dataclasses import dataclass
+from pprint import pprint
+from typing import Any, Callable, Optional
+
+import chip.clusters as Clusters
+import chip.tlv
+from chip import discovery
+from chip.clusters.Attribute import ValueDecodeFailure
+from chip.exceptions import ChipStackError
+from chip.setup_payload import SetupPayload
+from matter_testing_support import AttributePathLocation, MatterBaseTest, async_test_body, default_matter_test_main
+from mobly import asserts
+
+
+def MatterTlvToJson(tlv_data: dict[int, Any]) -> dict[str, Any]:
+    """Given TLV data for a specific cluster instance, convert to the Matter JSON format."""
+
+    matter_json_dict = {}
+
+    key_type_mappings = {
+        chip.tlv.uint: "UINT",
+        int: "INT",
+        bool: "BOOL",
+        list: "ARRAY",
+        dict: "STRUCT",
+        chip.tlv.float32: "FLOAT",
+        float: "DOUBLE",
+        bytes: "BYTES",
+        str: "STRING",
+        ValueDecodeFailure: "ERROR",
+        type(None): "NULL",
+    }
+
+    def ConvertValue(value) -> Any:
+        value_type = type(value)
+
+        if value_type == ValueDecodeFailure:
+            raise ValueError(f"Bad Value: {str(value)}")
+
+        if value_type == bytes:
+            return base64.b64encode(value).decode("UTF-8")
+        elif value_type == list:
+            for idx, item in enumerate(value):
+                value[idx] = ConvertValue(item)
+        elif value_type == dict:
+            value = MatterTlvToJson(value)
+
+        return value
+
+    for key in tlv_data:
+        value_type = type(tlv_data[key])
+        value = copy.deepcopy(tlv_data[key])
+
+        element_type: str = key_type_mappings[value_type]
+        sub_element_type = ""
+
+        new_key = ""
+        if element_type:
+            if sub_element_type:
+                new_key = f"{str(key)}:{element_type}-{sub_element_type}"
+            else:
+                new_key = f"{str(key)}:{element_type}"
+        else:
+            new_key = str(key)
+
+        try:
+            new_value = ConvertValue(value)
+        except ValueError as e:
+            new_value = str(e)
+
+        if element_type:
+            if element_type == "ARRAY":
+                if len(new_value):
+                    sub_element_type = key_type_mappings[type(tlv_data[key][0])]
+                else:
+                    sub_element_type = "?"
+
+        matter_json_dict[new_key] = new_value
+
+    return matter_json_dict
+
+
+def check_int_in_range(min_value: int, max_value: int, allow_null: bool = False) -> Callable:
```

Comment:
Should this perhaps take some string that describes where we came across this value, to be used in the error strings?

Then check_list_of_ints_in_range could use this checker, and pass in the "At index..." bit as a description.

Followup OK for this, or just not doing it.

Reply:
We can improve as we add more tests. I wanted to avoid over-optimizing error formats until there is more use.

Reply:
Like I said: followup is ok.